### PR TITLE
requirements for HTTP provisioning object

### PR DESCRIPTION
**Pull request by BertKleewein**
# Description of the problem
Initial SRS for device provisioning HTTP transport module

### CODE REVIEW

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+### constructor(idScope: string, registrationId: string) : void
+The `constructor` creates an Http transport object used to communicate with the Azure device provisioning service
+
+**SRS_NODE_PROVISIONING_HTTP_18_001: [** The `Http` constructor shall accept the following properties:
+- `idScope` - a string specifiying the scope of the provisioning operations,
+- `registrationId` - the registration id for the specific device 
+**]**
+
+
+### register(authorization?: string, forceRegistration?: boolean, callback (err: Error, result?: Any) => void) : void;
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >callback [](start = 66, length = 8)

missing ':'. also you can't have a mandatory parameter after optional parameters

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_004: [** The registration request shall use Http 1.1. **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_005: [** The registration request shall include the current `api-version` as a URL query string value. **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_006: [** The registration request shall specify the following in the Http header:
+  Accept: application/json
+  Content-Type: application/json; charset=utf-8 **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_007: [** If an `authorization` string is specifed, the registration request shall include this in the Http header. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >Http [](start = 134, length = 4)

maybe add "Http Authorization header" and the expected format?

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_008: [** If `forceRegistration` is specified, the registration request shall include this as a query string value. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >value [](start = 140, length = 5)

what's the query string parameter name?

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_009: [** `register` shall shall POST the registration request to 'https://global.azure-devices-provisioning.net/{idScope}/registrations/{registrationId}' **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >shall [](start = 58, length = 5)

repeated 'shall'

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_010: [** `register` shall wait for a registration response. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: This one is not super explicit - maybe rephrase saying when the callback should be called? once we're done polling and the operation succeeded, or when the response to the POST request comes in?

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_011: [** If the registration request times out, `register` shall call the `callback` with `TimeoutError` error **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_012: [** If the registration response contains a body, `register` shall deserialize this into a JSON object. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >JSON [](start = 128, length = 4)

Elio would say that JSON is a notation and by definition a string :) "into an object" is simple enough - and maybe needs a type?

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_013: [** If registration response body fails to deserialize, `register` will call the `callback` with a `SyntaxError` error. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >SyntaxError [](start = 137, length = 11)

- does this error already exist in the SDK? 
- shouldn't we throw here? means it's a programmer error somewhere, likely in the SDK? I don't want to give the customer the impression that this is something they are responsible for by giving them the error in a callback when in fact we are.

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_014: [** If the registration response has a failed status code, `register` shall use `translateError` to translate this to a common error object and pass this into the `callback` function along with the deserialized body of the response. **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_015: [** If the registration response has a success code with a 'status' of 'Assigned', `register` call the `callback` with `err` == `null` and result `containing` the deserialized body **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >`err` == `null` [](start = 156, length = 15)

I've learned recently that it's recommended to avoid using null when undefined can do. I'm torn between starting to use undefined and make this inconsistent with the rest of the SDK or keeping consistency (and to a certain extent, a bad practice) alive. not sure what to suggest but letting you know what's going on.

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_016: [** If the registration response has a success code with a 'status' of 'Assigning', `register` shall fire an `operationStatus` event with the deserialized body **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_017: [** If the registration response has a success code with a 'status' of 'Assigning', `register` shall start polling for operation updates **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_029: [** If the registration response has a success code with a 'status' that is not 'Assigned' or  'Assigning', `register` shall call the callback with a `SyntaxError` error. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >SyntaxError [](start = 188, length = 11)

same as the other syntax error

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+## Overview
+This module provides HTTP protocol support to communicate with the Azure device provisioning service
+
+```js
+'use strict';
+
+export class Http extends EventEmitter implements DeviceProvisioningTransport {
+  api-version: string;
+  operationPollingIntervalInSeconds: Number
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >operationPollingIntervalInSeconds [](start = 2, length = 33)

intervals are usually in milliseconds in Javascript, not sure what to do here.
Also, 'operation' is not super descriptive of what we're actually polling for.
Last but not least, what are other SDKs doing for APIs to set this?

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_018: [** `register` shall poll for operation status every `operationPollintervalInSeconds` seconds **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_019: [** The operation polling request shall use Http 1.1 **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_020: [** The operation polling request shall have the following in the Http header:
+  Accept: application/json
+  Content-Type: application/json; charset=utf-8 **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_021: [** If an `authorization` string is specifed, the operation polling request include it in the Http header. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >header [](start = 136, length = 6)

same: which header + expected format

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+**SRS_NODE_PROVISIONING_HTTP_18_022: [** operation polling shall be a GET operation sent to 'https://global.azure-devices-provisioning.net/{idScope}/registrations/{registrationId}/operations/{operationId}' **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_023: [** If the operation polling request times out, `register` shall stop polling and call the `callback` with a `TimeoutError` error. **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_024: [** `register` shall deserialize the body of the operation polling respone into a JSON object. **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_025: [** If the body of the operation polling respone fails to deserialize, `register` will call the `callback` with a `SyntaxError` error. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >SyntaxError [](start = 152, length = 11)

same as other SyntaxError comments

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
+# azure-iot-provisioning-http Requirements
+
+## Overview
+This module provides HTTP protocol support to communicate with the Azure device provisioning service
+
+```js
+'use strict';
+
+export class Http extends EventEmitter implements DeviceProvisioningTransport {
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >DeviceProvisioningTransport [](start = 50, length = 27)

would be interesting to see what this is.

Comment (BertKleewein) [reply provisioning/transport/http/devdoc/http_requirements.md]: I will actually flush this out when I have 2 protocols.  Right now, I'm assuming it will be register+disconnect+maybe some of the strings like api-version.

---
In reply to: [140063985](https://github.com/Azure/azure-iot-sdk-node/pull/86#discussion_r140063985) [](ancestors = 140063985)

File: provisioning/transport/http/devdoc/http_requirements.md (hunk 10 above)
Comment (BertKleewein) [reply provisioning/transport/http/devdoc/http_requirements.md]: I will compare to other SDKs shortly

---
In reply to: [140062992](https://github.com/Azure/azure-iot-sdk-node/pull/86#discussion_r140062992) [](ancestors = 140062992)

File: provisioning/transport/http/devdoc/http_requirements.md (hunk 6 above)
Comment (BertKleewein) [reply provisioning/transport/http/devdoc/http_requirements.md]: Good point.  I don't want to tie this to a type yet.  Even though the schema appears well defined at this point, I don't have a reason  to enforce that schema yet.

---
In reply to: [140060986](https://github.com/Azure/azure-iot-sdk-node/pull/86#discussion_r140060986) [](ancestors = 140060986)

File: provisioning/transport/http/devdoc/http_requirements.md (hunk 13 above)
Comment (pierreca) [reply provisioning/transport/http/devdoc/http_requirements.md]: sounds reasonable

---
In reply to: [140100915](https://github.com/Azure/azure-iot-sdk-node/pull/86#discussion_r140100915) [](ancestors = 140100915,140063985)

File: provisioning/transport/http/devdoc/http_requirements.md (hunk 10 above)
Comment (pierreca) [reply provisioning/transport/http/devdoc/http_requirements.md]: I suspect there will be a setOptions thing somewhere :-/

---
In reply to: [140107581](https://github.com/Azure/azure-iot-sdk-node/pull/86#discussion_r140107581) [](ancestors = 140107581,140062992)

File: provisioning/transport/http/devdoc/http_requirements.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  Accept: application/json
+  Content-Type: application/json; charset=utf-8 **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_007: [** If an `authorization` string is specifed, it shall be URL encoded and included in the Http Authorization header. **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_008: [** If `forceRegistration` is specified, the registration request shall include this as a query string value named 'forceRegistration' **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_009: [** `register` shall POST the registration request to 'https://global.azure-devices-provisioning.net/{idScope}/registrations/{registrationId}' **]**
+
+**SRS_NODE_PROVISIONING_HTTP_18_010: [** `register` shall wait for the response to the POST request. **]**
```
Comment (pierreca) [inline provisioning/transport/http/devdoc/http_requirements.md]: >[** `register` shall wait for the response to the POST request. **]** [](start = 37, length = 69)

so the idea is that register calls the callback after a successful POST request and then we use events to notify the user of progress, am I getting this right?  can the example usage reflect that?

File: provisioning/transport/http/devdoc/http_requirements.md (hunk 6 above)
Comment (pierreca) [reply provisioning/transport/http/devdoc/http_requirements.md]: "yet" nevertheless it'll make it easier for users to have autocompletion with at least an interface. even if for now you leave it as simple as a basic generic dictionnary, it will give us a clue not to forget it.

---
In reply to: [140107605](https://github.com/Azure/azure-iot-sdk-node/pull/86#discussion_r140107605) [](ancestors = 140107605,140060986)

File: provisioning/transport/http/devdoc/http_requirements.md (hunk 7 above)
Comment (pierreca) [reply provisioning/transport/http/devdoc/http_requirements.md]: not convinced by InternalServerError either. might be us screwing up somewhere. I'm not against creating a specific error for that but blaming it on the server doesn't seem right. throwing seems right though

---
In reply to: [140061844](https://github.com/Azure/azure-iot-sdk-node/pull/86#discussion_r140061844) [](ancestors = 140061844)

**Comment by pierreca** [comment]
That's a very good starting point although a few things will need to be ironed out as other transports come online :shipit:
